### PR TITLE
Some code style changes to meet PEP8

### PR DESCRIPTION
**Pull request by SR4ven**
The Flake8 check is currently outputting a large number of warnings (~400). Most of them are due to violations of [PEP8](https://www.python.org/dev/peps/pep-0008/).

The majority of those warnings are caused by deviating indentation or whitespaces. More complex ones that I didn't see a reason to fix or didn't know how to, are now being ignored. Search for `noqa` in the code and let me know if you disagree with any ignore.

Some init files were causing many `imported but unused` warnings. Adding them to the `__all__` variable fixes that ([reference](https://stackoverflow.com/questions/31079047/python-pep8-class-in-init-imported-but-not-used)). This also makes writing new protocol definitions easier when using an IDE with code completion, as blocks like `s_static()` are now being recognised.
I get that we need this for e.g. `s_initialize` so that it can be called independently, but do we need this for all loggers, blocks, constants and so on?

Additionally all imports are now sorted by system, package, local and alphabet (done with PyCharm).

I know this will be a lot of work to review, so take your time.

### CODE REVIEW

File: boofuzz/__init__.py [hunk 1]
```diff
@@ -1,37 +1,70 @@
...
-                         Byte, Word, DWord, QWord, FromFile, Mirror)
+from .primitives import (BasePrimitive, BitField, Byte, Delim, DWord, FromFile, Group, Mirror, QWord, RandomData,
+                         Static, String, Word)
 from .serial_connection import SerialConnection
-from .sessions import Session, Target, open_test_run
-from .exception import SullyRuntimeError, SizerNotUtilizedError, MustImplementException
+from .sessions import open_test_run, Session, Target
 from .socket_connection import SocketConnection
-from past.builtins import map
-from builtins import chr
```
Comment (SR4ven) [inline boofuzz/__init__.py]: Why do we import all submodules here? Do they all have to be callable independently without `boofuzz.`? Same question for other inits.

Comment (jtpereyda) [reply boofuzz/__init__.py]: The goal is similar to the goal described in the question here: https://stackoverflow.com/questions/31079047/python-pep8-class-in-init-imported-but-not-used

The idea is to enable `from boofuzz import Socketconnection` etc. instead of `from boofuzz.socket_connection import SocketConnection`.

File: boofuzz/blocks/block.py [hunk 2]
```diff
@@ -68,7 +67,7 @@ def original_value(self):
     def name(self):
         return self._name
 
-    def mutate(self):
+    def mutate(self):  # noqa: C901
```
Comment (jtpereyda) [inline boofuzz/blocks/block.py]: IMO "function is too complex" is a legitimate observation here. I'm working on another branch (mutations-iterator) on some changes that may end up helping.

File: unit_tests/test_size_original_value.py [hunk 3]
```diff
@@ -116,7 +116,7 @@ def result_equals_render_after_reset(context):
     assert context.result == context.uut.render()
 
 
-@then('Result equals .render() after Request reset()')
+@then('Result equals .render() after Request reset()')  # noqa: F811
```
Comment (jtpereyda) [inline unit_tests/test_size_original_value.py]: It's peculiar that these are required only on these two declarations. 🤷‍♂

File: boofuzz/__init__.py (hunk 1 above)
Comment (SR4ven) [reply boofuzz/__init__.py]: Makes sense, let's leave it as it is for now.

File: boofuzz/blocks/block.py (hunk 2 above)
Comment (SR4ven) [reply boofuzz/blocks/block.py]: Good to hear, I'll remove that ignore then.

Right now the max complexity is set to 15 and the mutate function is the only one with an even higher complexity.
The default would be 10, which in our case is violated by quite a few functions (e.g. sessions/init).

> This feature is quite useful to detect over-complex code. According to McCabe, anything that goes beyond 10 is too complex. See https://en.wikipedia.org/wiki/Cyclomatic_complexity.
https://pypi.org/project/flake8/2.4.1/

If you want to be more strict about the code complexity we can turn it down to the default, but it's going to be hard to follow imo.

Here is a list of all functions that violate the default complexity:
```python
./network_monitor.py:247:1: C901 'main' is too complex (12)
./vmcontrol.py:40:5: C901 'VMControlPedrpcServer.__init__' is too complex (15)
./vmcontrol.py:514:1: C901 'If 514' is too complex (15)
./utils/ida_fuzz_library_extender.py:90:1: C901 'find_strings' is too complex (11)
./boofuzz/sessions.py:323:5: C901 'Session.__init__' is too complex (12)
./boofuzz/sessions.py:1036:5: C901 'Session.transmit_normal' is too complex (15)
./boofuzz/sessions.py:1091:5: C901 'Session.transmit_fuzz' is too complex (13)
./boofuzz/socket_connection.py:119:5: C901 'SocketConnection.open' is too complex (15)
./boofuzz/socket_connection.py:176:5: C901 'SocketConnection.recv' is too complex (11)
./boofuzz/socket_connection.py:220:5: C901 'SocketConnection.send' is too complex (15)
./boofuzz/blocks/checksum.py:126:5: C901 'Checksum._checksum' is too complex (11)
./boofuzz/blocks/block.py:189:5: C901 'Block.render' is too complex (14)
```

File: boofuzz/helpers.py [hunk 4]
```diff
@@ -458,15 +455,15 @@ def get_boofuzz_version(boofuzz_class):
         for line in search:
             # line = line.rstrip()  # remove '\n' at end of line
             if line.find("__version__ = ") != -1:
-                version = 'v' + re.search("\'(.*?)\'", line).group(1)
+                version = 'v' + re.search(r"\'(.*?)\'", line).group(1)
```
[the remark this answers is not in the document]
Comment (SR4ven) [reply boofuzz/helpers.py]: It does indeed when using print. However the functions are working just as they did before.
The problem without treating those strings as raw is, that e.g `\s` is an invalid escape sequence (used in setup.py).
So we either define them as raw strings, or escape the characters like `\\s`.

In this exact case raw flag isn't necessary, I just added it because it's a regex and for consistency.
We could of course change the function in setup.py to work like this one, so the regex doesn't contain any invalid escape sequences anymore.

File: setup.py [hunk 5]
```diff
@@ -18,7 +18,7 @@ def find_version(*path_elements):
     """
     path = os.path.join(setup_dir, *path_elements)
     for line in open(path):
-        for match in re.finditer('__version__\s*=\s(.*)$', line):
+        for match in re.finditer(r"__version__\s*=\s(.*)$", line):
```
[the remark this answers is not in the document]
Comment (SR4ven) [reply setup.py]: The function still works as intended. `\s` is an invalid escape sequence, so there're 3 ways to handle this:
1. define string as raw
2. use `\\s`
3. use helpers.get_boofuzz_version() which doesn't contain any problematic characters

File: unit_tests/test_size_original_value.py (hunk 3 above)
Comment (SR4ven) [reply unit_tests/test_size_original_value.py]: My guess is that the check just runs from top to bottom and looks for redefinitions.
Would you rather have the ignore on both functions?
Or shall we fix this by simply renaming one of the functions?

File: setup.cfg [hunk 6]
```diff
@@ -3,3 +3,9 @@ zip_ok = False
 
 [zest.releaser]
 python-file-with-version = boofuzz/__init__.py
+
+[flake8]
+ignore = F403, F405, W503
```
[the remark this answers is not in the document]
Comment (SR4ven) [reply setup.cfg]: If we wouldn't ignore F403 and F405, we'd get ~1400 warnings. :)

We need [W503](https://lintlyci.github.io/Flake8Rules/rules/W503.html) here as it conflicts with W504. They are basically just inverted, but [W504](https://lintlyci.github.io/Flake8Rules/rules/W504.html) is the current one. No idea why they are both enabled by default.

File: boofuzz/primitives/string.py [hunk 7]
```diff
@@ -235,7 +236,7 @@ def __init__(self, value, size=-1, padding=six.binary_type(b"\x00"), encoding="a
         if self.max_len > 0:
             if any(len(s) > self.max_len for s in self.this_library):
                 # Pull out the bad string(s):
-                self.this_library = list(set([s[:self.max_len] for s in self.this_library]))
+                self.this_library = list(set([s[:self.max_len] for s in self.this_library]))  # noqa: F812
```
[the remark this answers is not in the document]
Comment (SR4ven) [reply boofuzz/primitives/string.py]: Would this be accurate?
```python
# Remove any fuzz items greater than self.max_len
if self.max_len > 0:
    if any(len(s) > self.max_len for s in self.this_library):
        # Pull out the bad string(s):
        self.this_library = list(set([t[:self.max_len] for t in self.this_library]))
    if any(len(s) > self.max_len for s in self._fuzz_library):
        # Pull out the bad string(s):
        self._fuzz_library = list(set([t[:self.max_len] for t in self._fuzz_library]))
```
It passes the test at least...

File: boofuzz/fuzz_logger_text.py [hunk 8]
```diff
@@ -22,7 +23,7 @@ class FuzzLoggerText(ifuzz_logger_backend.IFuzzLoggerBackend):
 
     def __init__(self, file_handle=sys.stdout, bytes_to_str=DEFAULT_HEX_TO_STR):
         """
-        :type file_handle: io.FileIO
+        :type file_handle: io.TextIO
```
[the remark this answers is not in the document]
Comment (SR4ven) [reply boofuzz/fuzz_logger_text.py]: The Print function isn't happy with the FileIO type in python3: `Expected type 'Optional[IO[str]]', got FileIO instead`

Using TextIO fixes that. The main difference between them two is that TextIO includes en-/decoding while FileIO is lower-level.
I'm not 100% sure if we really need FileIO here for logging the unformatted binary data, but we aren't really changing the type with this anyways. It's only the required type that an IDE might complain about.
`open('ftp-fuzz.csv', 'wb')` generates `_io.BufferedWriter` which is accepted by both TextIO and FileIO.

File: boofuzz/fuzz_logger_csv.py [hunk 9]
```diff
@@ -36,7 +36,7 @@ class FuzzLoggerCsv(ifuzz_logger_backend.IFuzzLoggerBackend):
     def __init__(self, file_handle=sys.stdout, bytes_to_str=DEFAULT_HEX_TO_STR):
         """
         Args:
-            file_handle (io.TextIOBase): Open file handle for logging. Defaults to sys.stdout.
+            file_handle (io.TextIO): Open file handle for logging. Defaults to sys.stdout.
```
[the remark this answers is not in the document]
Comment (SR4ven) [reply boofuzz/fuzz_logger_csv.py]: `open('ftp-fuzz.csv', 'wb')` generates `_io.BufferedWriter` (BinaryIO in pycharm) which matches only TextIO and [BinaryIO](https://docs.python.org/3/library/io.html#binary-i-o).
As we don't need en-/decoding here (?), I suggest we use BinaryIO instead.

FileIO doesn't match either of them so creating a file handle as above would cause a missleading type warning in an IDE.

[the remark this answers is not in the document]
Comment (jtpereyda) [reply boofuzz/fuzz_logger_csv.py]: I guess BinaryIO is appropriate, as the csv.writer docs say you should open files with the 'b' option: https://docs.python.org/2/library/csv.html#csv.writer

File: boofuzz/helpers.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (jtpereyda) [reply boofuzz/helpers.py]: Ah, right, since it's a regex, `\'` matches a literal `'` character, so there is no change in how the regex is parsed.

For simplicity, it would make sense to take out the two backslashes here. It won't change the meaning but will simplify the regex.

File: boofuzz/primitives/string.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (jtpereyda) [reply boofuzz/primitives/string.py]: Yes, that looks about right.

File: setup.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (jtpereyda) [reply setup.py]: OK, seems like it was incorrect before and the raw string corrects the issue. Great!

File: unit_tests/test_size_original_value.py (hunk 3 above)
Comment (jtpereyda) [reply unit_tests/test_size_original_value.py]: Oh, the function name is reused... Yes, simply renaming the functions seems like the intuitive fix.

This is easy to do while using the BDD style, since the function tag matters more than the function name. Having warnings enabled will hopefully make it easier to see when new warnings are added in the future.

File: setup.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (jtpereyda) [reply setup.py]: OK, I think I'm catching up now -- both versions are functionally correct, but the original version yields the flake8 warning `W605 invalid escape sequence '\s'`.

File: unit_tests/test_size_original_value.py (hunk 3 above)
Comment (SR4ven) [reply unit_tests/test_size_original_value.py]: I've renamed the functions but it does seem to have an effect on which of the two functions with the same name is called.
This doesn't change the outcome in the case of `result_equals_render_after_reset`, but does for `request_one_block`.
It appears that the redefinition overwrote `@given('A Size')` completely so that the test was always using the second function with `@given('A Size whose target block will change size upon mutation')`
Both function tags seem to point at the second, redefined function name.
As with the renaming it uses the intended function now, the unit tests fail because of an endless recursion.

I don't know how to quickly fix this so we either leave it as it was running the whole time by using the `@given('A Size whose target block will change size upon mutation')` for all scenarios or we fix the error it produces.
@jtpereyda feel free to add commits here to fix this!

Comment (SR4ven) [reply unit_tests/test_size_original_value.py]: @jtpereyda 0aafd5e might fix the problem.
I took a look at `request_one_block_with_size_change` and noticed, that `request.pop()` is called before defining the size block. So I applied the same structure on `request_one_block` and now the test passes.

I didn't really verify that the check does exactly what it's supposed to do, would be nice if you could double-check.

Comment (jtpereyda) [reply unit_tests/test_size_original_value.py]: @SR4ven I think your change is in line with the purpose of the test, as the text doesn't seem to describe any recursive behavior. I don't know if there's any unit test for Size under recursive conditions though, which is a bit of a gap. Seems like this change is good to go for now anyway.

**Comment by jtpereyda** [comment]
That was a lot of cleanup and a few minor bugs fixed. Thank you @SR4ven .
